source/common/stats/hystrix.h

```diff
@@ -0,0 +1,99 @@
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace Envoy {
+namespace Stats {
+
+typedef std::vector<uint64_t> RollingStats;
+typedef std::map<std::string, RollingStats> RollingStatsMap;
+
+// Consider implement the HystrixStats as a sink to have access to histograms data
```

Comment:
+1 to the idea of moving the bulk of the logic to a stats sink.

Reply:
We thought of doing this later as an enhancement..

Reply:
I was thinking more about using stats sinks for this.  I'd be interested to hear @mattklein123's and @htuch's thoughts.  IIUC, the general architecture is a subscription mechanism where a client reaches out to subscribe and then Envoy streams specific stats data indefinitely over that connection.  Is this correct?

That would mean that this is a combination of the existing pull-based integrations, like prometheus, that just implement a simple request-query-response handler on the admin port and push-based integrations, like the GRPC stats streamer and statsd, that implement stats sinks to push stats.  A lot of the custom functionality that you've added, like querying stats and periodic flushes, are already handled by the existing sink interface.  

I'm wondering if there's some way that we can keep an admin port subscription model, but pass off the stream to a stats sink where the periodic pushing can be done.  This would likely allow the pushing to be more performant since it won't require a new periodic query to be set up for each subscriber (assuming the multi-subscriber case would ever happen in a hystrix deployment).  It would also mean that you wouldn't have to reinvent the wheel when it comes to building and configuring a push system within Envoy or querying stats data.  And, as you mentioned, it will have built-in support for histograms.  I'm curious to hear what others think, though.

Reply:
Stats sink means the stream will be sent at same intervals as other stats which are active. Isn't it a limitation?

Reply:
@mattklein123, sounds good.  

@trabetti, yes, it is a bit of a limitation.  However, I think most deployments have only been using one sink so far, so the aligned intervals have not been an issue.  If this is something you foresee being an issue, we should open an issue and design a way for stats sinks to optionally specify a time interval that is different from the global one.  It shouldn't be too hard to make this more flexible.